Generate EventNames.h and EventNames.cpp from a JSON file

rniwa opened the pull request:
#### dc25bb5d4a2f1e9c8b1deed9546d6a8bdf38a6cf
<pre>
Generate EventNames.h and EventNames.cpp from a JSON file
<a href="https://bugs.webkit.org/show_bug.cgi?id=267125">https://bugs.webkit.org/show_bug.cgi?id=267125</a>

Reviewed by Elliott Williams, Tim Nguyen and David Kilzer.

This PR replaces EventNames.h with EventNames.json to generate EventNames.h and EventNames.cpp.

The newly introduced JSON file contains a dictionary of event names to their entries.
Each entry optionally specifies build condition and a set of types it belongs such as
&quot;Gamepad&quot; and &quot;TouchRelated&quot;.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreMacros.cmake:
* Source/WebCore/dom/EventNames.cpp: Removed.
* Source/WebCore/dom/EventNames.h: Removed.
* Source/WebCore/dom/EventNames.json: Added.
* Source/WebCore/dom/make-event-names.py: Added.
(lowercaseFirstLetter):
(main):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getSupportedEventNames):

Canonical link: <a href="https://commits.webkit.org/272705@main">https://commits.webkit.org/272705@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5bc7364e52169c897b5445a9e384d50db9683d87

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32766 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11517 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34626 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35339 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29595 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13863 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8702 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29071 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9717 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29267 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8462 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8601 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29219 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36676 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29774 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/29642 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34723 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8718 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6685 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32589 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10403 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9328 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4224 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9331 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer reviewed (approved):
r=me on the Makefile/build file changes.  The only thing I didn't look at closely was the Python code.

Looks like TV/Watch may need some conditionals adjusted to fix those builds.

ddkilzer commented on Source/WebCore/WebCoreMacros.cmake:
```diff
@@ -189,6 +189,19 @@ macro(GENERATE_EVENT_FACTORY _infile _namespace)
 
+macro(GENERATE_EVENT_NAMES _infile _namespace)
+    set(NAMES_GENERATOR ${WEBCORE_DIR}/dom/make-event-names.py)
+    set(_outputfiles ${WebCore_DERIVED_SOURCES_DIR}/${_namespace}EventNames.h ${WebCore_DERIVED_SOURCES_DIR}/${_namespace}EventNames.cpp)
```
I think the bug on CMake-based platforms is that this line needs to change because the generated code will create filenames of `EventEventNames.h` and `EventEventNames.cpp` the way it's currently written.

Maybe the `_namespace` argument isn't even needed for this code generation?

```suggestion
    set(_outputfiles ${WebCore_DERIVED_SOURCES_DIR}/${_namespace}Names.h ${WebCore_DERIVED_SOURCES_DIR}/${_namespace}Names.cpp)
```

rniwa replied:
Nice catch. Thanks.

nt1m commented on Source/WebCore/dom/EventNames.json:
```diff
@@ -0,0 +1,321 @@
+[
+    {"name": "DOMActivate", "types": ["MouseClickRelated"]},
```
No strong opinions, but it feels like this could be a key value object instead of an array, since the name is a required key ? Like `CSSProperties.json` or `CSSPseudoSelectors.json`.

rniwa replied:
Sure, changed.

nt1m commented on Source/WebCore/dom/make-event-names.py:
```diff
@@ -0,0 +1,204 @@
+import json
+
+
+def lowercaseFirstLetter(string):
```
nit: python style is usually `lowercase_first_letter`

rniwa replied:
Fixed.

nt1m commented on Source/WebCore/dom/make-event-names.py:
```diff
@@ -0,0 +1,204 @@
+        event_names_input = json.load(event_names_file)
+
+    with open('EventNames.h', 'w') as output_file:
+        def write(text):
+            output_file.write(text)
+
+        def writeln(text):
+            write(text + "\n")
```
If we can somehow share the writer utilities from CSSPseudoSelectors.json & CSSProperties.json (they're duplicated in both of those), you might be able to reuse those if you want to.

rniwa replied:
Hm... that would require quite a bit of refactoring so kept as is for now.

nt1m reviewed (commented):
This is pretty neat.

nt1m reviewed (approved):
LGTM, I'd just update the commit message since this is no longer an array of entries

emw-apple reviewed (approved):
Build system changes look reasonable to me.

nt1m commented on Source/WebCore/dom/make-event-names.py:
I noticed this file uses a mix of single and double quotes for strings, would be nice to make them consistent. My preference is single quotes, since it's consistent with other python files.

rniwa replied:
Oh oops, I missed this comment. Will address in a follow up.

rniwa committed:
Generate EventNames.h and EventNames.cpp from a JSON file
https://bugs.webkit.org/show_bug.cgi?id=267125

Reviewed by Elliott Williams, Tim Nguyen and David Kilzer.

This PR replaces EventNames.h with EventNames.json to generate EventNames.h and EventNames.cpp.

The newly introduced JSON file contains a dictionary of event names to their entries.
Each entry optionally specifies build condition and a set of types it belongs such as
"Gamepad" and "TouchRelated".

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/WebCoreMacros.cmake:
* Source/WebCore/dom/EventNames.cpp: Removed.
* Source/WebCore/dom/EventNames.h: Removed.
* Source/WebCore/dom/EventNames.json: Added.
* Source/WebCore/dom/make-event-names.py: Added.
(lowercaseFirstLetter):
(main):
* Source/WebCore/inspector/agents/InspectorDOMAgent.cpp:
(WebCore::InspectorDOMAgent::getSupportedEventNames):

Canonical link: https://commits.webkit.org/272705@main
